### PR TITLE
BUG: pd.offsets.MonthEnd() delevers wrong month

### LINKED ISSUE #48779: BUG: pd.offsets.MonthEnd() delevers wrong month
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
ts = pd.Timestamp(2017, 1, 31)
ts + pd.offsets.MonthEnd()
```

### Issue Description

I expect that the answer is: Timestamp('2017-01-31 00:00:00')
but I get: Timestamp('2017-02-28 00:00:00')

### Expected Behavior

Timestamp('2017-01-31 00:00:00')

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 4bfe3d07b4858144c219b9346329027024102ab6
python           : 3.9.12.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19044
machine          : AMD64
processor        : Intel64 Family 6 Model 140 Stepping 1, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : Dutch_Netherlands.1252

pandas           : 1.4.2
numpy            : 1.21.5
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.2.4
setuptools       : 61.2.0
Cython           : 0.29.28
pytest           : 7.1.1
hypothesis       : None
sphinx           : 4.4.0
blosc            : None
feather          : None
xlsxwriter       : 3.0.3
lxml.etree       : 4.8.0
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.3
IPython          : 8.2.0
pandas_datareader: None
bs4              : 4.11.1
bottleneck       : 1.3.4
brotli           : 
fastparquet      : None
fsspec           : 2022.02.0
gcsfs            : None
markupsafe       : 2.0.1
matplotlib       : 3.5.1
numba            : 0.55.1
numexpr          : 2.8.1
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.7.3
snappy           : 
sqlalchemy       : 1.4.32
tables           : 3.6.1
tabulate         : 0.8.9
xarray           : 0.20.1
xlrd             : 2.0.1
xlwt             : None
zstandard        : None

</details>

### PR DESCRIPTION
**Pull request by jtimko16**
[DOC] - Added more examples and explanation regarding MonthEnd method.

- [ ] closes #48779
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/_libs/tslibs/offsets.pyx [hunk 1]
```diff
@@ -2476,12 +2476,38 @@ cdef class MonthOffset(SingleConstructorOffset):
...
+    ----------
+    n : int, default 1
+    
     Examples
     --------
     >>> ts = pd.Timestamp(2022, 1, 1)
-    >>> ts + pd.offsets.MonthEnd()
+    >>> ts + pd.offsets.MonthEnd(n=1)
     Timestamp('2022-01-31 00:00:00')
+
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:2495]: can you add a short sentence before the examples you've added to clarify what the reader is meant to take away from them? see https://pandas.pydata.org/docs/development/contributing_docstring.html#section-7-examples

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
This is off to a good start - could you update the PR title to be the same as the issue?

**Comment by MarcoGorelli** [comment]
you'll also need
```
closes #48779
```
in the PR description (as per the template)

**Comment by MarcoGorelli** [comment]
and finally, you should check that `pre-commit run --files pandas/_libs/tslibs/offsets.pyx` passes

**Comment by honza26** [comment]
@MarcoGorelli Thanks for your comments. Very useful. I've made some updates now and another updates will follow this evening!

**Comment by honza26** [comment]
@MarcoGorelli  I still receive error regarding trailing white spaces:

image

However I removed all the white spaces in the end of lines. Not sure what else I could do. Do you have any idea, what may cause this problem?

**Comment by MarcoGorelli** [comment]
the hook has taken care of it for you - now you can just do `git add -u` and `git commit` again

as an aside, [please don't post images of code](https://youtu.be/svlsnZFWWDE)

File: pandas/_libs/tslibs/offsets.pyx [hunk 2]
```diff
@@ -2477,11 +2477,57 @@ cdef class MonthEnd(MonthOffset):
...
     DateOffset of one month end.
 
+    Parameters
+    ----------
+    n : int, default 1
+        n = 0 -> offset to the end of month.
+        n = 1 -> depend on the given date:
+            a) Given date is on an anchor point (last day of the month) -> offset to the end of the following month.
+            b) Given date is not on an anchor point -> offset to the end of the same month.
+
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:2487]: does this look right? can you try building this page of the docs to check please? https://pandas.pydata.org/docs/dev/development/contributing_documentation.html

File: pandas/_libs/tslibs/offsets.pyx [hunk 3]
```diff
@@ -2477,11 +2477,57 @@ cdef class MonthEnd(MonthOffset):
...
+    
+    However be careful, if given date is the last day of a month, method offsets to the end of the following month:
+    
+    >>> ts = pd.Timestamp(2022, 1, 31)
+    >>> ts + pd.offsets.MonthEnd()
+    Timestamp('2022-02-28 00:00:00')
+    
+    With the ``n`` parameter, we can control logic of offset.
+    
+    See below, when ``n`` is set to 0, offset is always made the end of the current month:
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:2504]:
```suggestion
    See below, when ``n`` is set to 0, offset is always made to the end of the current month:
```

File: pandas/_libs/tslibs/offsets.pyx [hunk 4]
```diff
@@ -2477,11 +2477,57 @@ cdef class MonthEnd(MonthOffset):
...
+    >>> ts + pd.offsets.MonthEnd(n=0)
+    Timestamp('2022-01-31 00:00:00')
+    
+    When ``n`` is set to 1 (default) it will be moved to the following month:
+    
+    >>> ts = pd.Timestamp(2022, 1, 31)
+    >>> ts + pd.offsets.MonthEnd(n=1)
+    Timestamp('2022-02-28 00:00:00')
+    
+    When ``n`` is set to 2, method offsets one additional month. And again depends whether given date is on on anchor point:
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:2516]:
```suggestion
    When ``n`` is set to 2, method offsets one additional month. And again depends whether given date is on an anchor point:
```

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

**Comment by honza26** [comment]
> the hook has taken care of it for you - now you can just do `git add -u` and `git commit` again
> 
> as an aside, [please don't post images of code](https://youtu.be/svlsnZFWWDE)

Ok, sure. Let me do it tomorrow. Today I am editing directly on the website, as I don't have access to my personal laptop to run git commands. I'll let you know tomorrow evening, in case the issue remains. Thanks.

And ok.

File: pandas/_libs/tslibs/offsets.pyx (hunk 2 above)
Comment (honza26) [reply pandas/_libs/tslibs/offsets.pyx:2487]: Ok, but I will need more time for this. I still have trouble with setting up virtual environment (spend one hour with it today and will continue tomorrow)

Comment (honza26) [reply pandas/_libs/tslibs/offsets.pyx:2487]: What is the version of Python you are using? I am created venv on top of Python 3.10, and trying to install package from requirements-dev.txt, but it always stuck at installation of boto3:
` 
  >>Using cached boto3-1.24.38-py3-none-any.whl (132 kB)
  >>Using cached boto3-1.24.37-py3-none-any.whl (132 kB)
  >>Using cached boto3-1.24.36-py3-none-any.whl (132 kB)
  Using cached boto3-1.24.35-py3-none-any.whl (132 kB)
  Using cached boto3-1.24.34-py3-none-any.whl (132 kB)
  Using cached boto3-1.24.33-py3-none-any.whl (132 kB)
  Using cached boto3-1.24.32-py3-none-any.whl (132 kB)`

Is it better to use older version of Python - e.g. 3.9? Thanks

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/offsets.pyx:2487]: I'd suggest using mamba https://pandas.pydata.org/docs/dev/development/contributing_environment.html#option-1a-using-mamba-recommended

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/offsets.pyx:2487]: please do ask (here or on Slack) if it's unclear / anything trips you up

Comment (honza26) [reply pandas/_libs/tslibs/offsets.pyx:2487]: Thank you so much. I managed to install and setup mamba environment. However I have the same issue as I had before when I created virtual environment in PyCharm (missing module 'pandas._libs.interval'):
 
` "C:\Programming\Python\pandas\pandas\_libs\__init__.py", line 13, in <module>
    from pandas._libs.interval import Interval
ModuleNotFoundError: No module named 'pandas._libs.interval'`

Do  you know what could be causing this issue? Thanks

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/offsets.pyx:2487]: yeah you'll need to rebuild your C extensions
```
python setup.py build_ext -j 4
```

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/offsets.pyx:2487]: this needs doing any time one of the .pyx or .c files changes, which could happen if you've just fetched and merge upstream/main

Comment (honza26) [reply pandas/_libs/tslibs/offsets.pyx:2487]: Thanks for a quick turnaround. However, when I run the commend you've suggested, I have another error:

`error: command 'cl.exe' failed: None`

I am doing my best to help, but I feel I am discovering more problems than solutions. Sorry!

Comment (honza26) [reply pandas/_libs/tslibs/offsets.pyx:2487]: Will look at it again tomorrow. Hopefully I will make better progress 🤞

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/offsets.pyx:2487]: no idea - could you show your complete traceback please?

Comment (honza26) [reply pandas/_libs/tslibs/offsets.pyx:2487]: I tried to run it again, and also installed C++ Build Tools, but still receiving similar error. Please see below the full traceback:
[Traceback_error.txt](https://github.com/pandas-dev/pandas/files/9833587/Traceback_error.txt)

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/offsets.pyx:2487]: I don't know, sorry - you could ask on [Slack ](https://pandas.pydata.org/docs/dev/development/community.html?highlight=slack#community-slack), or try using Ubuntu via the WSL on Windows

Comment (honza26) [reply pandas/_libs/tslibs/offsets.pyx:2487]: Hi, thank you. I don't want to sound like giving up. However unfortunately, this is going deeper and deeper technical, and I feel I am not able to effectively contribute. Is it OK, if I share the updated text of documentation in thread under the issue, and leave it for someone more technically capable to proceed with implementation?

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/offsets.pyx:2487]: sure, that's fine, someone else can take this forwards, and if you feel like trying again, feel free to reach out and ask

Comment (honza26) [reply pandas/_libs/tslibs/offsets.pyx:2487]: Ok, thanks. Maybe I will give it another go this weekend. Will let you know how I decide to proceed then. Thanks

**Comment by MarcoGorelli** [comment]
> This pull request is stale because it has been open for thirty days with no activity. Please [update](https://pandas.pydata.org/pandas-docs/stable/development/contributing.html#updating-your-pull-request) and respond to this comment if you're still interested in working on this.

closing for now, but feel free to ping if you want to keep working on this and I can reopen

**Pull request closed without merging**